python/src/iceberg/expressions/base.py

```diff
@@ -122,3 +125,139 @@ def __le__(self, other):
 
     def __ge__(self, other):
         return self.value >= other.value
+
+
+class BooleanExpression(ABC):
+    """base class for all boolean expressions"""
+
+    @abstractmethod
+    def __invert__(self) -> "BooleanExpression":
+        ...
+
+
+class And(BooleanExpression):
+    """AND operation expression - logical conjunction"""
+
+    def __new__(cls, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if rest:
+            return reduce(And, (left, right, *rest))
+        if left is AlwaysFalse() or right is AlwaysFalse():
+            return AlwaysFalse()
+        elif left is AlwaysTrue():
+            return right
+        elif right is AlwaysTrue():
+            return left
+        return super().__new__(cls)
+
+    def __init__(
+        self,
+        left: BooleanExpression,
+        right: BooleanExpression,
+        *rest: BooleanExpression,
+    ):
+        if not rest:
+            self.left = left
+            self.right = right
```

Comment:
This still need to be fixed. Even if we don't expect calls to `__init__` with both left/right and rest, we should not ignore the possibility.

Reply:
Hey Ryan, I was trying to suggest that it's not plainly ignored and that `__new__` takes care of making sure `__init__` *is* only called with with `left` and `right`. [This test](https://github.com/apache/iceberg/blob/0e637d74ee3899db8d0a6363c5abf31547377e0f/python/tests/expressions/test_expressions_base.py#L159) shows how `And` handles more arguments. Is that what you were thinking would not work?

Reply:
I think it's a great idea to handle this in `__new__`. But `__init__` still exists and can be called directly. I don't think it should assume it is always called through `__new__`. I think the only thing that you need to do is add an `else` case that throws `ValueError`.

Reply:
>  `__init__` still exists and can be called directly. I don't think it should assume it is always called through `__new__`. 

Would you mind elaborating on what you mean by this? To my knowledge, there is no calling init directly without implicitly calling new since new gives you the base object and init essentially adds/modifies the attributes

Reply:
> But __init__ still exists and can be called directly.

I don't think there's a way to do this while avoiding a call to `__new__`. The one way I wasn't sure of was if a child class calls init via super but I did this quick test and even that calls `__new__`.
```py
class Foo:
    def __new__(cls):
        print("New called")
        return cls
    def __init__(self, val):
        self.val = val

class Bar(Foo):
    def __init__(self):
        super().__init__(5)

Bar()  # "New called"
```

Reply:
I do think this could use some doc-strings describing this behavior though. Specifically mentioning __new__

Reply:
Calling `Foo.__init__(obj)` directly doesn't call `__new__`:

```python
f = Foo() # "New called"
Foo.__init__(f, 6) # avoids __new__
```

I'm not saying that these are likely cases, I'm just saying that we should prefer throwing exceptions for cases like this rather than failing silently.

```python
    def __init__(self, left, right, *rest):
        if not rest:
            self.left = left
            self.right = right
        else:
            raise ValueError("Expected only left and right operands")
```

It isn't that intrusive to have the `else` case and it will catch crazy cases because we can't prevent people from calling `__init__`.

Reply:
Looks like you can also avoid calling `__new__` by overriding it in the subclass:

```python
class Bar(Foo):
    def __new__(cls):
        print("Subclass new called")

Bar()  # "Subclass new called"
```

Reply:
I see, yeah I agree it's a small addition to protect against these cases. It also makes the init easier to understand.